Добавлена реализация системы рекомендаций

rabarcre opened the pull request.

rabarcre committed:
Добавлена реализация системы рекомендаций

Dvoychenkov commented on src/main/java/ru/yandex/practicum/filmorate/controller/RecommendationController.java:
```diff
@@ -0,0 +1,22 @@
+    private final RecommendationService recommendationService;
+
+    @GetMapping("/users/{id}/recommendations")
+    public List<Film> getRecommendations(@PathVariable Long id) throws SQLException {
```
На уровне контроллера экспешн SQLException не выкидываем, они у нас явно только в Row мапперах, например в UserRowMapper

Dvoychenkov commented on src/main/java/ru/yandex/practicum/filmorate/services/RecommendationService.java:
```diff
@@ -0,0 +1,30 @@
+    private final RecommendationStorage recommendationStorage;
+
+    public List<Film> getRecommendations(Long userId) {
+        checkId(userId);
```
Лучше проверку делать в стиле как например в UserService:

```
public Collection<UserDto> getFriends(Long userId) {
	getUserOrThrow(userId); // Проверка на наличие пользователя
	// Код по бизнес логике...
}

public User getUserOrThrow(Long id) {
	if (id == null) throw new ValidationException("Некорректный ID пользователя");
	User user = requireFound(userStorage.getById(id), () -> "Пользователь с ID " + id + " не найден");
	log.info("Получен пользователь по ID {}: {}", id, user);
	return user;
}
```

Dvoychenkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/recomendation/RecommendationDbStorage.java:
```diff
@@ -0,0 +1,48 @@
+    private final FilmRowMapper filmRowMapper;
+
+    @Override
+    public List<Film> getRecommendations(Long userId) {
```
Посмотрел сейчас на запрос и понял, что есть смысл указать вкратце комментом что в запросе происходит

Dvoychenkov reviewed (changes requested):
Есть что поправить для текущего варианта реализации

Dvoychenkov commented on src/main/java/ru/yandex/practicum/filmorate/services/RecommendationService.java:
```diff
@@ -0,0 +1,30 @@
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class RecommendationService {
```
Ещё момент для текущей реализации, мы условились делать интерфейс для новых сервисов, [вот пример](https://github.com/Dvoychenkov/java-filmorate/commit/e7203a88458ae409a8416adc2f0adb7369a085a8)

rabarcre committed:
Изменения связанные с ревью

Dvoychenkov committed:
- Удалено всё по теме отдельной сущности Recomendation - для такой задачи в этом нет необходимости;
- Дополнен интерфейс FilmStorage и добавлена реализация метода в FilmDbStorage + заглушка в InMemoryFilmStorage;
- Дополнен сервис и контроллер User (помещаем сюда, а не в Film классы согласно SRP);
- Удалено тело юнит-теста shouldNormalizeEmptyName, поскольку в develop ветке более актуальный вариант;

Dvoychenkov committed:
Merge branch 'develop' into add-recommendations

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/services/UserService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java
#	src/test/java/ru/yandex/practicum/filmorate/validation/UserValidationTest.java

Dvoychenkov committed:
- Фикс и рефакторинг юнит-теста;

Dvoychenkov committed:
- checkstyle fix;
